src/jimgw/transforms.py

```diff
@@ -0,0 +1,177 @@
+from abc import ABC, abstractmethod
+from typing import Callable
+
+import jax
+import jax.numpy as jnp
+from chex import assert_rank
+from jaxtyping import Float
+
+
+class Transform(ABC):
+    """
+    Base class for transform.
+
+    The idea of transform should be used on distribtuion,
+    """
+
+    name_mapping: tuple[list[str], list[str]]
+    transform_func: Callable[[dict[str, Float]], dict[str, Float]]
+
+    def __init__(
+        self,
+        name_mapping: tuple[list[str], list[str]],
+    ):
+        self.name_mapping = name_mapping
+
+    def __call__(self, x: dict[str, Float]) -> tuple[dict[str, Float], Float]:
+        return self.transform(x)
+
+    @abstractmethod
+    def transform(self, x: dict[str, Float]) -> tuple[dict[str, Float], Float]:
+        """
+        Transform the input x to transformed coordinate y and return the log Jacobian determinant.
+        This only works if the transform is a N -> N transform.
+
+        Parameters
+        ----------
+        x : dict[str, Float]
+                The input dictionary.
+
+        Returns
+        -------
+        y : dict[str, Float]
+                The transformed dictionary.
+        log_det : Float
+                The log Jacobian determinant.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def forward(self, x: dict[str, Float]) -> dict[str, Float]:
+        """
+        Push forward the input x to transformed coordinate y.
+
+        Parameters
+        ----------
+        x : dict[str, Float]
+                The input dictionary.
+
+        Returns
+        -------
+        y : dict[str, Float]
+                The transformed dictionary.
+        """
+        raise NotImplementedError
+
+    def propagate_name(self, x: list[str]) -> list[str]:
+        input_set = set(x)
+        from_set = set(self.name_mapping[0])
+        to_set = set(self.name_mapping[1])
+        return list(input_set - from_set | to_set)
+
+
+class UnivariateTransform(Transform):
+
+    def __init__(
+        self,
+        name_mapping: tuple[list[str], list[str]],
+    ):
+        super().__init__(name_mapping)
+
+    def transform(self, x: dict[str, Float]) -> tuple[dict[str, Float], Float]:
```

Comment:
Regarding transform and forward: my understanding is that forward only performs the map, while transform performs the map and computes the Jacobian as well. In that case, wouldn't it make sense to let the jax.jacfwd act on self.transform? Right now, there is some code duplication between transform and forward. 

Also, one could in principle get rid of using the self.transform_func and instead solely rely on self.transform if the desired transformations are written with dicts as input and output, right? For example:
```python
def Mc_q_to_m1_m2(params: dict):
    """Function taking in dict and transforming to another dict"""
    return_params = {}
    
    Mc = params['M_c']
    q = params['q']
    
    M = Mc * (1.0 + q) ** 1.2 / q**0.6
    m1 = M / (1.0 + q)
    m2 = M * q / (1.0 + q)
    
    return_params["m1"] = m1
    return_params["m2"] = m2
    
    return return_params
```

Reply:
I don't think calling `jax.jacfwd` on self.transform will work because that only returns is Jacobian. If you are thinking of calling one after another, popping the `name_mapping` of x may bug out since the program will pop x twice, causing instability. In general, the rationale behind having two separate functions, `transform` and `forward`, is that transform is more restrictive in terms of what `Transform` can be. `transform` can only be used when the kernel is n-to-n, where forward can be used in the case of n-to-m. In some cases, such as the prior transformation, we will want to use `transform` for the most part. In other case, such as interfacing with waveform generators, `forward` is preferred since it does not compute the Jacobian, which should be faster.

To me, the existence of `transform_func` is mostly for convenience, so the user doesn't have to take care of naming when they want to interact with the kernel of the function. On the other hand, `transform` and `forward` are designed with the purpose of interfacing with the rest of jim, so it carries the naming around.